### ISSUE TITLE
Hisense 100L5

### ISSUE THREAD

**Issue by chinedu40** [MrAsterisco/homebridge-hisense-tv#3]
**Describe The Bug:**

I have finally been able to get the device to authorise, please see NewAM thread about this. However the plugin doesn't work in home bridge and not working in the home app. I am able to successful send commands via command line e.g. python3.8 hisensetv.py --key home XX.XX.XX.XX --ifname XXX
**To Reproduce:**

**Expected behavior:**

For commands in home app/keyboard to work. 
**Logs:**

[4/15/2021, 6:40:16 PM] [HiSenseTV] An error occurred while fetching inputs: TypeError: Cannot read property 'join' of null
[4/15/2021, 6:40:26 PM] [HiSenseTV] An error occurred while fetching inputs: TypeError: Cannot read property 'join' of null
[4/15/2021, 6:40:36 PM] [HiSenseTV] An error occurred while fetching inputs: TypeError: Cannot read property 'join' of null
[4/15/2021, 6:40:46 PM] [HiSenseTV] An error occurred while fetching inputs: TypeError: Cannot read property 'join' of null
[4/15/2021, 6:40:56 PM] [HiSenseTV] An error occurred while fetching inputs: TypeError: Cannot read property 'join' of null
[4/15/2021, 6:41:06 PM] [HiSenseTV] An error occurred while fetching inputs: TypeError: Cannot read property 'join' of null
**Plugin Config:**

```json
Show your Homebridge config.json here, remove any sensitive information.
```
 {
            "ifname": "eth1",
            "devices": [
                {
                    "id": "Hisense 100L5FTUK",
                    "name": "Hisense PTV",
                    "ipaddress": "192.168.86.15",
                    "macaddress": "18:30:0c:c6:22:d8"
                }
            ],
            "platform": "HiSenseTV"

**Screenshots:**

**Environment:**

* **Plugin Version**:1.0
* **Homebridge Version**:  1.3.1
* **Node.js Version**: 
* **NPM Version**: 
* **Operating System**: Synology DSM 6.2

**Comment by MrAsterisco** [comment]
@chinedu40 thanks for reporting this! Could you please point me to the discussion on the other repo?

**Comment by chinedu40** [comment]
@MrAsterisco No problem, the thread is here https://github.com/newAM/hisensetv/issues/14

**Comment by chinedu40** [comment]
Just a slight correction to the above, i can switch on my projector using the plugin. I cannot turn it off however, it does not retrieve the inputs and i can't use the iOS remote function.

**Comment by MrAsterisco** [comment]
@chinedu40 Mmm... So, in the plugin, turning on is done via Wake On Lan, which is separated from the Python executable, so it makes sense.

**Comment by chinedu40** [comment]
So is it possible that the python executable code isn't running and hence would explain among other things why the plugin receive null when trying to run this section of code: 
this.sendCommand(['--get', 'sources'], (err, output) => {
      try {
        const response: InputSource[] = JSON.parse((output as any[]).join(''));
        this.inputSources = response;

        this.inputSources.forEach((inputSource, index) => {
          this.platform.log.debug('Adding input: ' + JSON.stringify(inputSource));

          const inputService = this.accessory.getService('input'+inputSource.sourceid) 
        || this.accessory.addService(this.platform.Service.InputSource, 'input'+inputSource.sourceid, 'input'+inputSource.sourceid);
        
          inputService.setCharacteristic(this.platform.Characteristic.IsConfigured, this.platform.Characteristic.IsConfigured.CONFIGURED);
          inputService.setCharacteristic(this.platform.Characteristic.ConfiguredName, inputSource.displayname);

          let inputType = this.platform.Characteristic.InputSourceType.OTHER;
          if (inputSource.sourcename === 'TV') {
            inputType = this.platform.Characteristic.InputSourceType.TUNER;
          } else if (inputSource.sourcename === 'AV') {
            inputType = this.platform.Characteristic.InputSourceType.COMPOSITE_VIDEO;
          } else if (inputSource.sourcename.startsWith('HDMI')) {
            inputType = this.platform.Characteristic.InputSourceType.HDMI;
          }
          
          inputService.setCharacteristic(this.platform.Characteristic.InputSourceType, inputType);
          inputService.setCharacteristic(this.platform.Characteristic.Identifier, (index+1));

          this.service.addLinkedService(inputService);
        });

        this.deviceState.hasFetchedInputs = true;
        this.getCurrentInput();
      } catch (error) {
        this.platform.log.error('An error occurred while fetching inputs: ' + error);
      }
    });

**Comment by chinedu40** [comment]
Okay, so i actually ran home bridge in debug mode which is new to me. This may be helpful. When trying to turn off the device  which uses the python execute this was the error:

[4/16/2021, 12:40:50 PM] [HiSenseTV] Set Characteristic On -> 0
[4/16/2021, 12:40:50 PM] [HiSenseTV] Sent power off, error: Error: ModuleNotFoundError: No module named 'paho'

**Comment by MrAsterisco** [comment]
Oh this is interesting. A quick Google search gave me this result: https://pypi.org/project/paho-mqtt/

It might be that you need to install that module in order to use the Python executable in that way? I didn't have to install it on my machine, but I'm also running Homebridge on Ubuntu, so it might have a different configuration.

**Comment by chinedu40** [comment]
What version of python is used in this plugin?

**Comment by MrAsterisco** [comment]
3.8, the same the plugin requires.

**Comment by chinedu40** [comment]
Interesting. So  i install paho-mqtt on my homebridge instance and now the error message has changed:

[4/16/2021, 1:02:53 PM] [HiSenseTV] Set Characteristic On -> 0
[4/16/2021, 1:02:54 PM] [HiSenseTV] Sent power off, error: Error: FileNotFoundError: [Errno 2] No such file or directory

**Comment by MrAsterisco** [comment]
😳  But if you manually run the Python executable embedded in the Homebridge plugin... does it work?

**Comment by chinedu40** [comment]
Yes running  python3.8 /volume1/docker/homebridge/node_modules/homebridge-hisense-tv-remotenow/bin/hisensetv.py --get sources XX:XX:XX:XX --ifname XXX works. 

A quick search on stackoverflow says that this error occurs when 'You are using a relative path, which means that the program looks for the file in the working directory. The error is telling you that there is no file of that name in the working directory.
Try using the exact, or absolute, path.'

So i wonder is it to do with this piece of code in platformAccessory.ts:  
sendCommand(args: string[], callback: (err?: Error, output?: any) => void) {
PythonShell.run(path.resolve(__dirname, '../bin/hisensetv.py'), {args: args.concat([this.accessory.context.device.ipaddress, '--ifname', this.platform.config.ifname])}, callback);
  }

**Comment by MrAsterisco** [comment]
Mmm... It could be, but I can't really use an absolute path there, because I would have to know where `node_modules` are installed in your system... Right?

**Comment by chinedu40** [comment]
This is true. I suppose the question is what directory is platformAccessory.ts supposed to be in the live plugin?

**Comment by MrAsterisco** [comment]
I'd say in the same directory as it is in the project right now, I'm not aware of any policy that moves stuff around in NPM packages. But I believe the issue is when launching the Python script, right? So it doesn't really matter where the `ts` file is, I'm afraid.

**Comment by chinedu40** [comment]
So after hours of trying to figure this out, i'm pleased to report i have done it!

Turns out this error: [4/16/2021, 1:02:53 PM] [HiSenseTV] Set Characteristic On -> 0
[4/16/2021, 1:02:54 PM] [HiSenseTV] Sent power off, error: Error: FileNotFoundError: [Errno 2] No such file or directory

Was caused by my initial change to __main__.py where i had to amend the line starting with add ssl_context = ssl._create_unverified_context to the following:
**add ssl_context = ssl._create_unverified_context(purpose=ssl.Purpose.CLIENT_AUTH)
ssl_context.load_cert_chain(certfile='/path/to/rcm_certchain_pem.cer', keyfile='/path/to/rcm_pem_privkey.pkcs8')**

because my device required particular .cer and .pkcs8 files to do successful tls authentication

I initially put the paths to these files somewhere other than where the homebridge plugin was looking eg /volume1/homes/chinedu40/path/to/files. I therefore had to change the above to

ssl_context.load_cert_chain(certfile='/homebridge/node_modules/homebridge-hisense-tv-remotenow/path/to/rcm_certchain_pem.cer', keyfile='/homebridge/node_modules/homebridge-hisense-tv-remotenow/path/to/rcm_pem_privkey.pkcs8')

For reference the .cer and .pkcs8 where obtained from here https://github.com/d3nd3/Hisense-mqtt-keyfiles.

**Comment by MrAsterisco** [comment]
This sounds awesome, thanks for putting so much effort into this! 
I'd be happy to review a PR to add support for this, since it sounds like you already fixed the problem!

**Comment by ryanshand** [comment]
> This sounds awesome, thanks for putting so much effort into this!
> I'd be happy to review a PR to add support for this, since it sounds like you already fixed the problem!

@MrAsterisco I created a new branch, uploaded the certificates and the edited main.py file into selfsignedcerts branch and created a pull request. 
My first time so I hope I did it right.

**Comment by ryanshand** [comment]
> This sounds awesome, thanks for putting so much effort into this!
> I'd be happy to review a PR to add support for this, since it sounds like you already fixed the problem!

OK @MrAsterisco I think this is ready to go. @chinedu40 your changes failed on my system as the path did not exist.

**Comment by MrAsterisco** [comment]
@ryanshand I left a comment on the PR, I am concerned about the static path... But seems you already solved that and I missed the conversation?

**Comment by ryanshand** [comment]
> @ryanshand I left a comment on the PR, I am concerned about the static path... But seems you already solved that and I missed the conversation?

Hey @MrAsterisco , I'm not fond of the static path either. I'm not great with Python - I tried for ages to set a dynamic path to the certs but when I tested it via Terminal, it kept timing out when connecting to the tv.
I tried this:

baseDir = os.getcwd()
certDir = os.path.join(baseDir, "Hisensecerts")
certFilePath = str(certDir + '/rcm_certchain_pem.cer')
keyFilePath= str(certDir + '/rcm_pem_privkey.pkcs8')

and then in the ssl context ....


if args.no_ssl:
     ssl_context = None
    else:
       ssl_context = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH)
       ssl_context.load_cert_chain(certfile=certFilePath, keyfile=keyFilePath)

when i test by adding print(keyFilePath) and print (certFilePath) it prints to terminal with the correct location
which for me on a Rasperry Pi "Bunker" running homebridge via npm is:

baseDir = 
/usr/local/lib/node_modules/homebridge-hisense-tv-remotenow/bin
certDir = 
'/usr/local/lib/node_modules/homebridge-hisense-tv-remotenow/bin/Hisensecerts

I don't know why this doesn't work but I spent about 8 hours on it.
In the end, it didn't matter what combination I tried, the ssl_context.load_cert_chain() only seemed to accept a static path.

Note: I can define the string as a variable as pass it in, which is what my PR has done - but if I use system commands to set the name, that doesn't seem to pass in well.

The reason I wanted to use the method above ^ is because that would allow it to function across any platform, as long as there was a directory called Hisensecerts inside the /bin folder and it contained the certificates.

Totally open to some feedback and/or help in getting that working, but it's got me stumped.

**Comment by MrAsterisco** [comment]
@ryanshand Understood. Thanks for the detailed explanation!

I have some experience with Python, but not with the SSL library, unfortunately... And I also can't figure out why it wouldn't work. What about passing the paths as a parameter to the script?

_BTW, I just realized that the changes only affect the Python script, so once we merge the PR, we should also bring the changes to the original fork. I can take care of that later on._

**Comment by ryanshand** [comment]
> @ryanshand Understood. Thanks for the detailed explanation!
> 
> I have some experience with Python, but not with the SSL library, unfortunately... And I also can't figure out why it wouldn't work. What about passing the paths as a parameter to the script?
> 
> _BTW, I just realized that the changes only affect the Python script, so once we merge the PR, we should also bring the changes to the original fork. I can take care of that later on._

I'm happy to give it a go, passing it in as a parameter - sorry I don't know how to do that - but if you could and let me know, I can download the files to my box here and give them a shot?

**Comment by MrAsterisco** [comment]
You can take a look at the `main.py` file, specifically the `def main()` function.

It should just be a matter of doing:

```python
parser.add_argument(
  "--ssl-cert", // or whatever name you feel appropriate
  type=str,
  help="A description of the parameter",
  default=""
)
```

and then you can retrieve the value later on by calling `args.ssl_cert`. You can then pass that to the constructor of `HisenseTv` (as we're doing already for other parameters on line [`113`](https://github.com/MrAsterisco/homebridge-hisense-tv/blob/b808410a326a5c1e91338796e3ec812008561bca/bin/hisensetv/__main__.py#L113)).

I can also do it, but it won't happen earlier than this weekend.

**Comment by ryanshand** [comment]
we can't pass it in as a parameter to the HisenseTV constructor, because it's part of the ssl context. we need to pass it into ssl_context.load_cert_chain(certFile,keyFile) as those 2 parameters.

I've tried 2 things now that fail:

I added to arguments

```
parser.add_argument("--certfile", type=str, help="location of certfile", default="")
parser.add_argument("--keyfile", type=str, help="location of certfile", default ="")

    args = parser.parse_args()
    
    sslparam1 = args.certfile
    sslparam2 = args.keyfile
    
    print(sslparam1)
    print(sslparam2)
```

you can see in the screenshot below, the print command shows the correct path to the file, and it definitely exists here.

![image](https://user-images.githubusercontent.com/83274883/116377590-ea2c9e80-a843-11eb-9546-8f44c7fefc3d.png)

2. I created 2 functions to define and return the path

```
def getCertsKey():
	
    baseDir = os.getcwd()
    certDir = os.path.join(baseDir, "Hisensecerts")
    certFilePath = str(certDir + '/rcm_certchain_pem.cer')
	
    return certFilePath
	
def getKeyFile():

    baseDir = os.getcwd()
    certDir = os.path.join(baseDir, "Hisensecerts")
    keyFilePath= str(certDir + '/rcm_pem_privkey.pkcs8')
	
    return keyFilePath
```


then in the main() function, I call these and store them

```
    sslparam1 = getCertsKey()
    sslparam2 = getKeyFile()
    
    print(sslparam1)
    print(sslparam2)
```

again, this shows the correct path when printed to screen, but times out when connecting to the TV.


![image](https://user-images.githubusercontent.com/83274883/116377935-3d065600-a844-11eb-8ce7-c44f1455e5cc.png)

**Comment by MrAsterisco** [comment]
Wow, this is really interesting. Thanks so much for putting the effort, but... Yeah, there seems to be no solution. I'll approve the PR as it is.

As mentioned earlier, I won't be able to publish a new version right now, but I'll try to do it asap.

Again, thank you so much for your help and for the huge amount of work you put into this!

**Comment by chinedu40** [comment]
Remember that if you're using a device that has already been authorized the tv will timeout.

**Comment by chinedu40** [comment]
> we can't pass it in as a parameter to the HisenseTV constructor, because it's part of the ssl context. we need to pass it into ssl_context.load_cert_chain(certFile,keyFile) as those 2 parameters.
> 
> I've tried 2 things now that fail:
> 
> I added two arguments
> 
> ```
> parser.add_argument("--certfile", type=str, help="location of certfile", default="")
> parser.add_argument("--keyfile", type=str, help="location of certfile", default ="")
> 
>     args = parser.parse_args()
>     
>     sslparam1 = args.certfile
>     sslparam2 = args.keyfile
>     
>     print(sslparam1)
>     print(sslparam2)
> ```
> 
> you can see in the screenshot below, the print command shows the correct path to the file, and it definitely exists here.
> 
> ![image](https://user-images.githubusercontent.com/83274883/116377590-ea2c9e80-a843-11eb-9546-8f44c7fefc3d.png)
> 
> 1. I created 2 functions to define and return the path
> 
> ```
> def getCertsKey():
> 	
>     baseDir = os.getcwd()
>     certDir = os.path.join(baseDir, "Hisensecerts")
>     certFilePath = str(certDir + '/rcm_certchain_pem.cer')
> 	
>     return certFilePath
> 	
> def getKeyFile():
> 
>     baseDir = os.getcwd()
>     certDir = os.path.join(baseDir, "Hisensecerts")
>     keyFilePath= str(certDir + '/rcm_pem_privkey.pkcs8')
> 	
>     return keyFilePath
> ```
> 
> then in the main() function, I call these and store them
> 
> ```
>     sslparam1 = getCertsKey()
>     sslparam2 = getKeyFile()
>     
>     print(sslparam1)
>     print(sslparam2)
> ```
> 
> again, this shows the correct path when printed to screen, but times out when connecting to the TV.
> 
> ![image](https://user-images.githubusercontent.com/83274883/116377935-3d065600-a844-11eb-8ce7-c44f1455e5cc.png)
> 
> It is worth noting that MQTT appears to connect (briefly) - when i set logging to DEBUG, I get a JSON response with what's playing on TV right now, but none of the other data comes back like AV inputs etc. So the issue isn't the cert location in and of itself (e.g. the error isn't that it can't find the file)

I see no reason why the first method wouldn't work, as mentioned above, try passing another argument other than authorize

**Comment by ryanshand** [comment]
> Remember that if you're using a device that has already been authorized the tv will timeout.
> 
> Try passing the --get state command instead using your method.

hmmm. this is --get state

![image](https://user-images.githubusercontent.com/83274883/116379315-78555480-a845-11eb-9654-292b7ada4085.png)


But if I hardcode the path again, it prompts me for a pin - so... I don't think you are correct regarding this authorize vs get state

![image](https://user-images.githubusercontent.com/83274883/116379872-04677c00-a846-11eb-969d-095fb2be5729.png)

**Comment by ryanshand** [comment]
@chinedu40 I have created a branch here https://github.com/ryanshand/homebridge-hisense-tv-ssl/tree/Addargs-SSL that has the changes I describe above.
Please feel free to test on your machine as well - if you compare ^ this version, and then again try hardcoded absolute path I am interested to see if you get the same outputs as me.

**Comment by ryanshand** [comment]
@MrAsterisco just checking will this rep be also updated?

**Comment by MrAsterisco** [comment]
@ryanshand Yes, sorry about the delay here, guys. I had a personal issue I had to attend to during the weekend. I merged the other PR and will now take care of updating this one. I hope to get to it today or tomorrow.

**Comment by ryanshand** [comment]
No rush, I was just double checking I hadn’t missed something.

**Comment by MrAsterisco** [comment]
Ah no! You've done an amazing job, so far! Thank you!

**Comment by MrAsterisco** [comment]
@ryanshand @chinedu40 This should all be included and (hopefully) working in version 1.1.0.

Please, update in your Homebridge and make sure to change the config so that it uses the expected `sslmode`. See the README for further details.

Feel free to reopen this issue if you encounter any issue!

_I also added both your names to the Credits section of this repo, hope you're ok with it._
